CRM/Utils/SQL/TempTable.php

```diff
@@ -0,0 +1,266 @@
+<?php
+/*
+ +--------------------------------------------------------------------+
+ | CiviCRM version 5                                                  |
+ +--------------------------------------------------------------------+
+ | Copyright CiviCRM LLC (c) 2004-2018                                |
+ +--------------------------------------------------------------------+
+ | This file is a part of CiviCRM.                                    |
+ |                                                                    |
+ | CiviCRM is free software; you can copy, modify, and distribute it  |
+ | under the terms of the GNU Affero General Public License           |
+ | Version 3, 19 November 2007 and the CiviCRM Licensing Exception.   |
+ |                                                                    |
+ | CiviCRM is distributed in the hope that it will be useful, but     |
+ | WITHOUT ANY WARRANTY; without even the implied warranty of         |
+ | MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.               |
+ | See the GNU Affero General Public License for more details.        |
+ |                                                                    |
+ | You should have received a copy of the GNU Affero General Public   |
+ | License and the CiviCRM Licensing Exception along                  |
+ | with this program; if not, contact CiviCRM LLC                     |
+ | at info[AT]civicrm[DOT]org. If you have questions about the        |
+ | GNU Affero General Public License or the licensing of CiviCRM,     |
+ | see the CiviCRM license FAQ at http://civicrm.org/licensing        |
+ +--------------------------------------------------------------------+
+ */
+
+/**
+ *
+ * @package CRM
+ * @copyright CiviCRM LLC (c) 2004-2018
+ *
+ * Table naming rules:
+ *   - MySQL imposes a 64 char limit.
+ *   - All temp tables start with "civicrm_tmp".
+ *   - Durable temp tables: "civicrm_tmp_d_{12}_{32}"
+ *   - Ephemeral temp tables: "civicrm_tmp_e_{12}_{32}"
+ *
+ * Example 1: Just create table name. You'll be responsible for CREATE/DROP actions.
```

Comment:
Given that the most common syntax above seems to be setCategory the omission of setCategory (& possibly setId) from these examples seems worth fixing

Reply:
Agree we should give an example of that in the docblock, so I'll add it.

But I'd also keep the existing examples. The `setCategory()` and `setId()` should be  optional -- most cases should work without them. But I did use them in refactoring because... I'm conservative. To wit: if someone previously called `DAO::createTempTableName()` and set the equivalent options, then I suppose they may have had a reason, and imitating them reduces our risk.

Digging in a little:

* With the predecessor `createTempTableName()`, about ~70% (9/13) specified a prefix (equiv to setting `category`) and ~7% (1/13) specified an id.
* The main benefit of setting a category is for debugging -- it makes it easier to read a query.
* The main benefit of setting an ID is reproducibility. Usually, we don't care if temptables have reproducible IDs (and we often explicitly don't want them to be reproduced). In the six cases I've updated, the only example which smelled like it mattered was `CRM/Contact/Form/Task.php`. (But I didn't really test -- I just kept the old ID to be safe.)